File: app/src/main/java/org/helpapaw/helpapaw/signaldetails/SignalDetailsPresenter.java
package org.helpapaw.helpapaw.signaldetails;

import org.helpapaw.helpapaw.base.Presenter;
import org.helpapaw.helpapaw.data.models.Comment;
import org.helpapaw.helpapaw.data.models.Signal;
import org.helpapaw.helpapaw.data.repositories.CommentRepository;
import org.helpapaw.helpapaw.data.repositories.PhotoRepository;
import org.helpapaw.helpapaw.data.repositories.SignalRepository;
import org.helpapaw.helpapaw.data.user.UserManager;
import org.helpapaw.helpapaw.utils.Injection;
import org.helpapaw.helpapaw.utils.Utils;

import java.util.Collections;
import java.util.List;

/**
 * Created by iliyan on 7/25/16
 */
public class SignalDetailsPresenter extends Presenter<SignalDetailsContract.View> implements SignalDetailsContract.UserActionsListener {

    private boolean showProgressBar;
    private List<Comment> commentList;
    private Signal signal;

    private CommentRepository commentRepository;
    private PhotoRepository photoRepository;
    private SignalRepository signalRepository;
    private UserManager userManager;

    public SignalDetailsPresenter(SignalDetailsContract.View view) {
        super(view);
        showProgressBar = true;
        commentRepository = Injection.getCommentRepositoryInstance();
        photoRepository = Injection.getPhotoRepositoryInstance();
        userManager = Injection.getUserManagerInstance();
        signalRepository = Injection.getSignalRepositoryInstance();
    }

    @Override
    public void onInitDetailsScreen(Signal signal) {
        setProgressIndicator(showProgressBar);
        if (signal != null) {
            this.signal = signal;
            signal.setPhotoUrl(photoRepository.getPhotoUrl(signal.getId()));
            getView().showSignalDetails(signal);

            if (commentList != null) {
                setProgressIndicator(false);

                if (commentList.size() == 0) {
                    getView().setNoCommentsTextVisibility(true);
                } else {
                    getView().displayComments(commentList);
                    getView().setNoCommentsTextVisibility(false);
                }
            } else {
                loadCommentsForSignal(signal.getId());
            }
        }
    }

    private void loadCommentsForSignal(String signalId) {
        commentRepository.getAllCommentsBySignalId(signalId, new CommentRepository.LoadCommentsCallback() {
            @Override
            public void onCommentsLoaded(List<Comment> comments) {
                if (!isViewAvailable()) return;
                commentList = comments;
                setProgressIndicator(false);

                if (commentList.size() == 0) {
                    getView().setNoCommentsTextVisibility(true);
                } else {
                    getView().displayComments(comments);
                    getView().setNoCommentsTextVisibility(false);
                }
            }

            @Override
            public void onCommentsFailure(String message) {
                if (!isViewAvailable()) return;
                getView().showMessage(message);
            }
        });
    }

    @Override
    public void onAddCommentButtonClicked(final String comment) {
        if (comment != null && comment.trim().length() > 0) {
            getView().hideKeyboard();
            setProgressIndicator(true);
            getView().scrollToBottom();

            userManager.isLoggedIn(new UserManager.LoginCallback() {
                @Override
                public void onLoginSuccess() {
                    if (!isViewAvailable()) return;
                    getView().clearSendCommentView();
                    saveComment(comment, signal.getId());
                }

                @Override
                public void onLoginFailure(String message) {
                    if (!isViewAvailable()) return;
                    setProgressIndicator(false);
                    getView().openLoginScreen();
                }
            });

        } else {
            getView().showCommentErrorMessage();
        }
    }

    @Override
    public void onStatusChanged(int status) {
        if(Utils.getInstance().hasNetworkConnection()) {
            signalRepository.updateSignalStatus(signal.getId(), status, new SignalRepository.UpdateStatusCallback() {
                @Override
                public void onStatusUpdated() {

                }

                @Override
                public void onStatusFailure(String message) {
                    if(!isViewAvailable()) return;
                    getView().showMessage(message);
                }
            });
        } else {
            getView().showNoInternetMessage();
        }
    }

    @Override
    public void onCallButtonClicked() {
        String phoneNumber = signal.getAuthorPhone();
        getView().openNumberDialer(phoneNumber);
    }

    private void saveComment(String comment, String signalId) {
        commentRepository.saveComment(comment, signalId, new CommentRepository.SaveCommentCallback() {
            @Override
            public void onCommentSaved(Comment comment) {
                if (!isViewAvailable()) return;
                setProgressIndicator(false);
                commentList.add(comment);
                getView().setNoCommentsTextVisibility(false);
                getView().displayComments(Collections.singletonList(comment));
            }

            @Override
            public void onCommentFailure(String message) {
                if (!isViewAvailable()) return;
                getView().showMessage(message);
            }
        });
    }

    private boolean isViewAvailable() {
        return getView() != null && getView().isActive();
    }

    private void setProgressIndicator(boolean active) {
        getView().setProgressIndicator(active);
        this.showProgressBar = active;
    }
}

File: app/src/main/java/org/helpapaw/helpapaw/signaldetails/SignalDetailsPresenter.kt
package org.helpapaw.helpapaw.signaldetails

import org.helpapaw.helpapaw.base.Presenter
import org.helpapaw.helpapaw.data.models.Comment
import org.helpapaw.helpapaw.data.models.Signal
import org.helpapaw.helpapaw.data.models.backendless.repositories.*
import org.helpapaw.helpapaw.data.user.BackendlessUserManager
import org.helpapaw.helpapaw.data.user.UserManager
import org.helpapaw.helpapaw.utils.Utils
import javax.inject.Inject


class SignalDetailsPresenter(override var view: SignalDetailsContract.View?) : Presenter<SignalDetailsContract.View>(view), SignalDetailsContract.UserActionsListener {

    @Inject
    constructor(signalDetailsFragment: SignalDetailsFragment):this(signalDetailsFragment as SignalDetailsContract.View)

    private var showProgressBar: Boolean = false
    private var commentList: MutableList<Comment>? = null
    lateinit var signal: Signal

    private var statusChanged: Boolean = false

    @Inject
    lateinit var utils: Utils

    @Inject
    lateinit var commentRepository: CommentRepository

    @Inject
    lateinit var photoRepository: BackendlessPhotoRepository

    @Inject
    lateinit var signalRepository: BackendlessSignalRepository

    @Inject
    lateinit var userManager: BackendlessUserManager

    init{
        showProgressBar = true
        statusChanged = false
    }


    override fun onInitDetailsScreen(signal: Signal?) {
        setProgressIndicator(showProgressBar)
        if (signal != null) {
            this.signal = signal
            signal.photoUrl = (photoRepository.getPhotoUrl(signal.id))
            view?.showSignalDetails(signal)

            if (commentList != null) {
                setProgressIndicator(false)

                if (commentList?.size == 0) {
                    view?.setNoCommentsTextVisibility(true)
                } else {
                    view?.displayComments(commentList)
                    view?.setNoCommentsTextVisibility(false)
                }
            } else {
                loadCommentsForSignal(signal.id)
            }
        }
    }

    override fun loadCommentsForSignal(signalId: String) {
        if (utils.hasNetworkConnection()) {
            commentRepository.getAllCommentsBySignalId(signalId, object : CommentRepository.LoadCommentsCallback {
                override fun onCommentsLoaded(comments: List<Comment>) {
                    if (!isViewAvailable()) return
                    commentList = comments.toMutableList()
                    setProgressIndicator(false)

                    if (commentList?.size == 0) {
                        view!!.setNoCommentsTextVisibility(true)
                    } else {
                        view!!.displayComments(comments)
                        view!!.setNoCommentsTextVisibility(false)
                    }
                }

                override fun onCommentsFailure(message: String) {
                    if (!isViewAvailable()) return
                    view!!.showMessage(message)
                }
            })
        } else {
            if (isViewAvailable()) {
                view!!.showNoInternetMessage()
            }
        }
    }

    override fun onAddCommentButtonClicked(comment: String?) {
        if (utils.hasNetworkConnection()) {
            if (comment != null && comment.trim().isNotEmpty()) {
                view!!.hideKeyboard()
                setProgressIndicator(true)
                view!!.scrollToBottom()

                userManager.isLoggedIn(object : UserManager.LoginCallback {
                    override fun onLoginSuccess() {
                        if (!isViewAvailable()) return
                        view!!.clearSendCommentView()
                        saveComment(comment, signal.id)
                    }

                    override fun onLoginFailure(message: String) {
                        if (!isViewAvailable()) return
                        setProgressIndicator(false)
                        view!!.openLoginScreen()
                    }
                })

            } else {
                view!!.showCommentErrorMessage()
            }
        } else {
            view!!.showNoInternetMessage()
        }
    }

    override fun onRequestStatusChange(status: Int) {
        if (utils.hasNetworkConnection()) {

            userManager.isLoggedIn(object : UserManager.LoginCallback {
                override fun onLoginSuccess() {
                    signalRepository.updateSignalStatus(signal.id, status, object : SignalRepository.UpdateStatusCallback {
                        override fun onStatusUpdated(status: Int) {
                            if (!isViewAvailable()) return
                            setSignalStatus(status)
                            view!!.showStatusUpdatedMessage()
                            view!!.onStatusChangeRequestFinished(true, status)
                        }

                        override fun onStatusFailure(message: String) {
                            if (!isViewAvailable()) return
                            view!!.showMessage(message)
                            view!!.onStatusChangeRequestFinished(false, 0)
                        }
                    })
                }

                override fun onLoginFailure(message: String) {
                    if (!isViewAvailable()) return
                    view!!.onStatusChangeRequestFinished(false, 0)
                    view!!.openLoginScreen()
                }
            })
        } else {
            view!!.showNoInternetMessage()
        }
    }

    override fun onCallButtonClicked() {
        val phoneNumber = signal.authorPhone
        view?.openNumberDialer(phoneNumber)
    }

    override fun onSignalDetailsClosing() {
        view?.closeScreenWithResult(signal)
    }

    override fun onBottomReached(isBottomReached: Boolean) {
        view?.setShadowVisibility(!isBottomReached)
    }

    override fun onSignalPhotoClicked() {
        view?.openSignalPhotoScreen()
    }

    private fun saveComment(comment: String, signalId: String) {
        commentRepository?.saveComment(comment, signalId, object : CommentRepository.SaveCommentCallback {
            override fun onCommentSaved(comment: Comment) {
                if (!isViewAvailable()) return
                setProgressIndicator(false)
                commentList?.add(comment)
                view!!.setNoCommentsTextVisibility(false)
                view!!.displayComments(commentList)
            }

            override fun onCommentFailure(message: String) {
                if (!isViewAvailable()) return
                view!!.showMessage(message)
            }
        })
    }


    private fun setSignalStatus(status: Int) {
        this.signal.status = status
        statusChanged = true
    }

    private fun isViewAvailable(): Boolean {
        return view != null && view!!.isActive()
    }

    private fun setProgressIndicator(active: Boolean) {
        view!!.setProgressIndicator(active)
        this.showProgressBar = active
    }
}